sweeper/sweeper.go

```diff
@@ -0,0 +1,191 @@
+package sweeper
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/skynetlabs/pinner/database"
+	"github.com/skynetlabs/pinner/skyd"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/build"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+type (
+	// Schedule defines how often, if at all, we sweep this server automatically.
+	Schedule struct {
+		Period   time.Duration
+		cancelCh chan struct{}
+	}
+	// Status represents the status of a sweep.
+	Status struct {
+		InProgress bool
+		Error      error
+		StartTime  time.Time
+		EndTime    time.Time
+	}
+	// Sweeper takes care of sweeping the files pinned by the local skyd server
+	// and marks them as pinned by the local server.
+	Sweeper struct {
+		staticDB         *database.DB
+		staticSkydClient skyd.Client
+		staticServerName string
+
+		schedule   *Schedule
+		scheduleMu sync.Mutex
+
+		status   Status
+		statusMu sync.Mutex
+	}
+)
+
+// New returns a new Sweeper.
+func New(db *database.DB, skydc skyd.Client, serverName string) *Sweeper {
+	return &Sweeper{
+		staticDB:         db,
+		staticSkydClient: skydc,
+		staticServerName: serverName,
+	}
+}
+
+// Schedule a scan to run on each period.
+func (s *Sweeper) Schedule(period time.Duration) {
+	s.scheduleMu.Lock()
+	defer s.scheduleMu.Unlock()
+	if s.schedule != nil {
+		close(s.schedule.cancelCh)
+	}
+	s.schedule = &Schedule{
+		Period:   period,
+		cancelCh: make(chan struct{}),
+	}
+	go func() {
+		t := time.NewTicker(s.schedule.Period)
+		for {
+			select {
+			case <-t.C:
+				s.Sweep()
+			case <-s.schedule.cancelCh:
+				return
+			}
+		}
+	}()
+}
+
+// Scheduled returns a copy of the current Sweeper schedule.
+func (s *Sweeper) Scheduled() Schedule {
+	if s.schedule == nil {
+		return Schedule{}
+	}
+	return *s.schedule
+}
+
+// Status returns a copy of the status of the current sweep.
+func (s *Sweeper) Status() Status {
+	s.statusMu.Lock()
+	st := s.status
+	s.statusMu.Unlock()
+	return st
+}
+
+// Sweep starts a new skyd sweep, unless one is already underway.
+//
+// TODO If we want to be able to uniquely identify sweeps we can issue ids
+//  for them and keep their statuses in a map. This would be the appropriate
+//  RESTful approach. I am not sure we need that because all we care about
+//  is to be able to kick off one and wait for it to end and this
+//  implementations is sufficient for that.
+func (s *Sweeper) Sweep() {
+	go s.threadedPerformSweep()
+}
+
+// threadedPerformSweep performs the actual sweep operation.
+func (s *Sweeper) threadedPerformSweep() {
+	s.statusMu.Lock()
+	// Double-check for parallel sweeps.
+	if s.status.InProgress {
+		s.statusMu.Unlock()
+		return
+	}
+	// Initialise the status to "a sweep is running".
+	s.status = Status{
```

Comment:
this could be extracted into a constructor `newStatus` and the whole thing could be `callInitSweep` on the status that returns a `bool` indicating if one started, in case it returns `false` you could return

Reply:
I went with the Start/Finalize approach and a static status. Otherwise I would have needed two mutexes to protect a single field and that would have been confusing. Now Start initialises a sweep (unless one is already in progress) and Finalize finishes it.

Reply:
ok